lib/LLVMIRCodeGen/JITFilePrinter.cpp

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2020 by Cadence Design Systems, Inc. ALL RIGHTS RESERVED.
+ *
+ * These coded instructions, statements, and computer programs are the
+ * copyrighted works and confidential proprietary information of Cadence Design
+ * Systems Inc. They may be adapted and modified by bona fide purchasers for
+ * internal use, but neither the original nor any adapted or modified version
+ * may be disclosed or distributed to third parties in any manner, medium, or
+ * form, in whole or in part, without the prior written consent of Cadence
+ * Design Systems Inc or in accordance with the terms of your Design Technology
+ * License Agreement with Cadence Design Systems. This software and its
+ * derivatives are to be executed solely on products incorporating a Cadence
+ * Design Systems processor.
+ */
+
+// This file contains the print functionality the JIT code can use. The backend
+// can call function to emit a global var that points to the host side struct
+// containing fopen/fprintf/fclose pointers, JIT code can call instead of it's
+// own fopen/fprintf/fclose.
+
+#include "glow/LLVMIRCodeGen/JITFilePrinter.h"
+#include "glow/LLVMIRCodeGen/LLVMIRGen.h"
+
+using namespace glow;
+
+#if defined _MSC_VER
+#define EXTERN __declspec(dllexport)
+#else
+#define EXTERN
+#endif
+
+extern "C" {
+static JITFileWriter jitfileWriter;
+}
+
+int JITFileWriter::JITfprintf(FILE *f, const char *format, ...) {
+  va_list argptr;
+  va_start(argptr, format);
+  return vfprintf(f, format, argptr);
+}
+
+FILE *JITFileWriter::JITfopen(const char *path, const char *mode) {
+  // File access is locked using per file name lock. This lock is a simple
+  // bool variable with it's access guarded by mutex.
+  std::unique_lock<std::mutex> lck(jitfileWriter.mtx_);
+  jitfileWriter.readyCV.wait(
+      lck, [&path] { return !jitfileWriter.fileLock_[std::string(path)]; });
+  jitfileWriter.fileLock_[std::string(path)] = true;
+  FILE *f = fopen(path, mode);
+  jitfileWriter.fileMap_[f] = std::string(path);
+  return f;
+}
+
+int JITFileWriter::JITfclose(FILE *f) {
+  std::lock_guard<std::mutex> lck(jitfileWriter.mtx_);
+  std::string file = jitfileWriter.fileMap_[f];
+  int ret = fclose(f);
+  jitfileWriter.fileLock_[file] = false;
+  jitfileWriter.readyCV.notify_all();
+  return ret;
+}
+
+void LLVMIRGen::generateJITFileWriter(llvm::Function *F) {
```

Comment:
Please add comments explaining what this function does and how it is implemented.